stablehlo/reference/Element.cpp

```diff
@@ -350,6 +350,33 @@ Element min(const Element &e1, const Element &e2) {
       });
 }
 
+// These cases are implementation-defined, and the behavior of this interpreter:
+// For integers: if rhs is 0, the result is the same as `lhs`.
+//   For signed integers, if lhs is INT_MAX and rhs is -1, the result is 0.
+// For floats: use APFloat.mod(). If rhs is 0, the result is 0x7FF8000000000000
+Element rem(const Element &e1, const Element &e2) {
+  return map(
+      e1, e2,
+      [&](APInt lhs, APInt rhs) {
+        if (rhs == 0) return lhs;
+        return isSupportedSignedIntegerType(e1.getType()) ? lhs.srem(rhs)
+                                                          : lhs.urem(rhs);
+      },
+      [](bool lhs, bool rhs) -> bool {
+        llvm::report_fatal_error("reminder of two bool numbers is unsupported");
+      },
+      [](APFloat lhs, APFloat rhs) {
+        lhs.mod(rhs);
```

Comment:
Why did we choose `APFloat::mod` over `APFloat::remainder`?

Reply:
Note that: APFloat::remainder(17.0, 3.0) = -1.000000e+00

Let's see the code: 
APFloat::mod is using [IEEEFloat::mod](https://llvm.org/doxygen/classllvm_1_1detail_1_1IEEEFloat.html#af98a9f1655f7320272c11d555e7e6846) while APFloat::remainder is using [IEEEFloat::remainder](https://llvm.org/doxygen/classllvm_1_1detail_1_1IEEEFloat.html#a878514b4bc5399e8ba0b104aa81d2fdf)

Our spec is: *For floating-point element types, this operation is in contrast with the remainder operation from IEEE-754 specification where d is an integral value nearest to the exact value of lhs/rhs with ties to even.*

I think the spec does not have a preference? (Please correct me if wrong). So considering the consistence that rem(17, 3) is 2, we should have rem(17.0, 3.0) as 2.0 instead of -1.0. 

Btw, HLO is using C mod, same as `APFloat::mod` above. https://github.com/tensorflow/tensorflow/blob/c4c3c066d48983f57afd9e8fbe3734b03ee98018/tensorflow/compiler/xla/hlo/evaluator/hlo_evaluator_typed_visitor.h#L700

What do you think of this?

Reply:
Thank you for explaining! Can you call out in a comment that using mod and not remainder is deliberate?

Reply:
Sure! Added.